### ISSUE TITLE
SolutionDescriber issues

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/molarity#67]
- [x] Missing documentation of overview and responsibilities:

```js
/**
 * @author Michael Kauzmann (PhET Interactive Simulations)
 * @author Taylor Want (PhET Interactive Simulations)
 */
```

- [x] Difficult to tell because of missing documentation overview/responsibilities.  But `SolutionDescriber` seems to be responsible for more than just the solution. E.g. `getStateOfSimDescription`, `getBeakerDescription`, `getSliderAlertString` .

- [x] Constructor creates unnecessary coupling:

```js
    /**
     * @param {MolarityModel} model
     * @param {BooleanProperty} valuesVisibleProperty - tracks whether the "Show values" checkbox is checked
     */
    constructor( model, valuesVisibleProperty ) {

      // @private
      this.model = model;
      this.solution = model.solution;
      this.valuesVisibleProperty = valuesVisibleProperty;
...
```

There is no need for SolutionDescriber to know about the entire model, and `this.model` is in fact not used.   Change the constructor to:

```js
    /**
     * @param {Solution} solution
     * @param {BooleanProperty} valuesVisibleProperty - tracks whether the "Show values" checkbox is checked
     */
    constructor( solution, valuesVisibleProperty )
```

More generally, in the [code_review_checklist.md](https://github.com/phetsims/phet-info/blob/master/checklists/code_review_checklist.md):

> - [ ]  Is there any unnecessary coupling? (e.g., by passing large objects to constructors, or exposing unnecessary properties/functions)

Passing in the entire model is an example of unnecessary coupling.  Unless it results in an unreasonable number of parameters, it's best to parameterize a constructor/method with only the information that is necessary.    This prevents the constructor/method from having access to things that it should not be accessing.

- [x] There's no need to add functions `soluteAmountToIndex` and `volumeToIndex` to the class definition.  They are private and use none of the class' state, so they can be private functions that are defined outside the class definition.  `concentrationToIndex` could also be private if you added parameter `saturatedConcentration`. The PhET convention is to put such functions after the class definition. WebStorm complains about this, since it's generally a good pattern to follow.

- [x] `getAlertString` has parameter `{string} slider ` that indicates which slider has been moved.  There's no indication of what the valid values are, and no assertion to validate the argument value, which is required for string params used like enumerations.   And from the [Enumeration section of phet-software-design-patterns.md](https://github.com/phetsims/phet-info/blob/master/doc/phet-software-design-patterns.md#enumerations), using strings as enumerations "should be avoided" in new code. 

- [x] This bit of code appears in `getStateOfSimDescription`, and the "Show Values" is _NOT_ initially  checked when this sim starts up.  Is this just a confusing comment? Or is there really some assumption here about initial state.

```js
      // sets values to fill in initially -- "Show Values" checkbox is checked
      let saturatedConcentration = '';
      let concentrationPattern = '';
```

- [x] Multiple lines are longer than 120 characters.

- [x] Inadequate documentation for `getQuantities`; "object with values from model"  is not very informative.

**Comment by pixelzoom** [comment]
Same problem with `MolarityAlertManager` constructor.  `model` is unnecessary.  And `this.solutions` does not need to be a field, since it's used only in the constructor.

**Comment by pixelzoom** [comment]
- [ ] `getAlertString` has parameter `{string} slider ` that indicates which slider has been moved.  There's no indication of what the values are, and no assertion to validate the value, which is required for string params used like enumerations.   And from the [Enumeration section of phet-software-design-patterns.md](https://github.com/phetsims/phet-info/blob/master/doc/phet-software-design-patterns.md#enumerations), using strings as enumerations "should be avoided" in new code.

**Comment by pixelzoom** [comment]
@zepumph fyi.

**Comment by twant** [comment]
@pixelzoom I have changed SolutionDescriber quite a bit based on conversations with @zepumph , and I believe those revisions have addressed most of your review comments. I am still planning to refactor a bit for readability. 

The only comment we still need to think about is whether to rename SolutionDescriber and its documentation to indicate that it is responsible for more descriptions than just the solution's, or whether to create additional describers to reduce the responsibilities of SolutionDescriber. Right now, since most of the descriptions are still related to the solution, I'm leaning toward a more general MolarityDescriber, which is responsible for all descriptions and alert text, but open to other ideas!

Let me know if you have other thoughts, and thanks again for your review comments!

**Comment by pixelzoom** [comment]
SolutionDescriber is failing CT, see https://github.com/phetsims/molarity/issues/72.  Assign back to me when that's fixed.

**Comment by twant** [comment]
@pixelzoom just pushed quite a few changes (apologies for the large commit), including a complete restructuring of the describers. I broke out the SolutionDescriber into a series of much smaller describers (for each of the model properties). I also removed MolarityAlertManager, and moved the property links to MolarityScreenView. Let me know if you see any issues, or have ideas for further improvements!

**Comment by pixelzoom** [comment]
This separation into multiple Describes looks much better.  But it looks like they are combined into MolarityDescriber.   Can you describe what pattern is being used here? 

Also note the commits above, let me know if you have any questions about these.  There were quite a few incorrect/missing type expressions, incorrectly described parameters, etc.

**Comment by pixelzoom** [comment]
I also suggest that you go through all of the documentation again.  It looks like there are copy-paste errors.  For example, in ConcentrationDescriber, I see this identical documentation for 3 methods:

```js
    /**
     * Gets the saturated concentration amount of the currently selected solute.
     * @private
     * @returns {number}
     */
```

**Comment by twant** [comment]
@pixelzoom Thanks for this feedback, and apologies for the formatting/documentation issues -- I did another pass this week to clean up formatting and documentation in the describers.

Regarding the describer pattern, when I broke the describer up into multiple smaller describers, it became clear that many (if not most) of the descriptions require information about all four of the properties with their own describers (volume, solute amount, solute, and concentration). Molarity Describer is intended to be the overarching structure that puts all of those smaller descriptive pieces into longer, more complex descriptions. I'll discuss ways to clean this up with @zepumph this week, but if you have thoughts on better ways to structure this, I'd love your input!

**Comment by pixelzoom** [comment]
Looking good.  Let me know when you'd like me to review.

**Comment by twant** [comment]
@pixelzoom @zepumph it seems that the describers have evolved quite a bit since we left off with this issue. Okay with you if we close this issue and review the describer structure in #112?

**Comment by zepumph** [comment]
The only checkbox I see left in this issue is:

>Is there any unnecessary coupling? (e.g., by passing large objects to constructors, or exposing unnecessary properties/functions)

I don't see this anymore. Ready to close.
